Sources/PenguinStructures/AnyArrayBuffer.swift

```diff
@@ -12,127 +12,121 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-/// A resizable, value-semantic buffer of homogenous elements of
-/// statically-unknown type.
-public struct AnyArrayBuffer<Storage: AnyArrayStorage> {
-  /// A bounded contiguous buffer comprising all of `self`'s storage.
-  ///
-  /// Note: `storage` has reference semantics. Clients that mutate the `storage` must take care to
-  /// preserve `ArrayBuffer`'s value semantics by ensuring that `storage` is uniquely referenced.
-  public var storage: Storage
-  
-  init(storage: Storage) { self.storage = storage }
-  
-  public init<SrcStorage>(_ src: ArrayBuffer<SrcStorage>) {
-    // The downcast would be unnecessary but for
-    // https://bugs.swift.org/browse/SR-12906
-    self.storage = unsafeDowncast(src.storage, to: Storage.self)
-  }
+/// Dynamic dispatchers for AnyArrayBuffer operations.
+// TODO: Add an example.
+public protocol AnyArrayDispatch: AnyObject { associatedtype Element }
 
-  /// Creates a buffer with elements from `src`.
+/// Helpers for concrete implementations
+extension AnyArrayDispatch {
+  /// Returns the array storage whose address is `p`.
   ///
-  /// Precondition: `SrcStorage: Storage` (we could express this in the
-  /// signature but for https://bugs.swift.org/browse/SR-12906).
-  public init<SrcStorage>(_ src: AnyArrayBuffer<SrcStorage>) {
-    self.storage = unsafeDowncast(src.storage, to: Storage.self)
+  /// - Requires: `p` is the address of an initialized `ArrayStorage<Element>`.
+  public static func asStorage(_ p: UnsafeRawPointer) -> ArrayStorage<Element> {
+    assert(
+      type(of: p.assumingMemoryBound(to: AnyObject.self).pointee)
+        == ArrayStorage<Element>.self,
+    "Expected storage type mismatch"
+    )
+    return p.assumingMemoryBound(to: ArrayStorage<Element>.self).pointee
   }
+}
 
-  /// Returns a buffer with elements from `self`, and storage type
-  /// `DesiredStorage`, if `Self.Storage` can be cast to `DesiredStorage`.
-  public func cast<DesiredStorage>(to _: DesiredStorage.Type)
-    -> AnyArrayBuffer<DesiredStorage>?
-  {
-    guard let desiredStorage = storage as? DesiredStorage else { return nil }
-    return AnyArrayBuffer<DesiredStorage>(storage: desiredStorage)
+extension AnyArrayBuffer where Dispatch == Void {
```

Comment:
Can we use `AnyObject` for buffers without operations instead of `Void`, so that we can cast buffers with operations to `AnyArrayBuffer<AnyObject>`?

Reply:
That sounds brilliant and obvious; wish I'd thought of it.  Feel free to make the change.